console-plugins/dense-mapping-constraints-plugin/src/dense-mapping-selection.cc

```diff
@@ -116,6 +128,86 @@ bool selectAlignmentCandidatePairs(
           << candidate_pairs_ptr->size() << " based on " << num_good_prior_edges
           << " good prior constraints and removed " << num_removed_edges
           << " bad prior constraints.";
+}
+
+static void filter_candidates_randomly(
+    const std::size_t max_number_of_candidates,
+    const std::size_t n_recent_candidates,
+    AlignmentCandidatePairs* candidate_pairs_ptr) {
+  CHECK_NOTNULL(candidate_pairs_ptr);
+
+  // Create a vector of candidate iterators.
+  const std::size_t n_candidates = candidate_pairs_ptr->size();
+  std::vector<AlignmentCandidatePairs::iterator> v(n_candidates);
+  std::iota(v.begin(), v.end(), candidate_pairs_ptr->begin());
+
+  // Sorter for the iterators based on the newest timestamp.
+  auto sorter = [](const AlignmentCandidatePairs::iterator& lhs,
+                   const AlignmentCandidatePairs::iterator& rhs) {
+    const int64_t newest_lhs_ts_ns = lhs->getNewestTimestamp();
+    const int64_t newest_rhs_ts_ns = rhs->getNewestTimestamp();
+    return newest_lhs_ts_ns > newest_rhs_ts_ns;
+  };
+
+  // Sort the iterators to access the most recent ones.
+  // Shuffle the remaining iterators.
+  std::sort(v.begin(), v.end(), sorter);
+  std::shuffle(
+      v.begin() + n_recent_candidates, v.end(),
+      std::mt19937{std::random_device{}()});
+
+  // Delete the elements from the original candidate list.
+  const std::size_t n_remaining_candidates = n_candidates - n_recent_candidates;
+  const std::size_t n_candidates_to_delete =
+      n_remaining_candidates -
+      std::min(n_remaining_candidates, max_number_of_candidates);
+  auto it = v.begin() + n_recent_candidates;
+  const auto it_end = it + n_candidates_to_delete;
+  for (; it != it_end; ++it) {
+    candidate_pairs_ptr->erase(*it);
+  }
+}
+
+static void filter_candidates_based_on_strategy(
+    const SelectionConfig& config, vi_map::VIMap* map_ptr,
+    AlignmentCandidatePairs* candidate_pairs_ptr) {
+  if (config.max_number_of_candidates == 0) {
+    return;
+  }
+  CHECK_NOTNULL(candidate_pairs_ptr);
+  CHECK_NOTNULL(map_ptr);
+
+  // Compute the bounds of the number of candidates to retrieve and reserve.
+  const std::size_t n_recent_candidates = static_cast<std::size_t>(
+      static_cast<double>(config.max_number_of_candidates) *
+      config.prioritize_recent_candidates);
+  const std::size_t n_max_candidates =
```

Comment:
could this be calculated internally in `filter_candidates_randomly` or named differently? It's a bit counterintuitive that max_candidates has a different meaning and value in different places

Reply:
I would probably keep it outside and name it differently. It could be that other filter strategies rely on the same information.